### ISSUE TITLE
Type inference with `Union[list[T], T]` as a parameter

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#2642]
Probably related to #2568 ?

Code:
```py
from typing import TypeVar, Union, Sequence

T = TypeVar("T")

def to_list(t: Union[list[T], T]) -> SequenceT]:
    if isinstance(t, list):
        return t
    return [t]

x = foo([1, 2, 3])
reveal_type(x)
```
`x` is inferred to be `Sequence[list[int]]`. This makes sense because `T` is assigned to be `list[int]`. But at runtime `x` will actually be `list[int]`.

So the bigger question is: should the signature of `to_list` be considered valid? Consider this function:
```py
def foo(x: T):
    bar = to_list(x)
    reveal_type(bar)
```
`bar` is revealed to be a `list[T]`. But if `T` is `list[int]`, then `bar` should theoretically be something like `Sequence[E] where issubclass(T, list[E])` which is currently not expressable.

**Comment by erictraut** [comment]
This isn't related to the other issue.

In cases where multiple solutions are possible within the TypeVar constraint solver, pyright attempts to apply some heuristics to select the "simplest" and "most complete" solution (i.e., provides values for the most type variables). In the example above, there are two possible solutions for `T`: `int` and `list[int]`. The simpler of the two solutions is `int`, which is the expected result, but the heuristics didn't select that one. I've improved the heuristics so the algorithm does a better job at selecting the simplest solution.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.192, which I just published. It will also be included in the next release of pylance.
